### PR TITLE
Extract fingerprinting for experiment and model config

### PR DESCRIPTION
**Pull request by frgossen**
Extract fingerprinting, into a function, which was anonymously implemented in two places before. This is used to resume benchmarking from where it stopped. Configurations are identified by their fingerprint.

Clearly distinguish between the subprocess and the main process in the code, trying to make it more readable.

### CODE REVIEW

File: benchmarks/util.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ def reset_rng_state(benchmark_experiment=None):
   torch.manual_seed(1337)
   random.seed(1337)
   np.random.seed(1337)
-  if benchmark_experiment and benchmark_experiment.xla:
+  if benchmark_experiment is not None and benchmark_experiment.xla:
```
Comment (cota) [inline benchmarks/util.py:44]: `if foo` is the same as `if foo is not None`, and AFAICT is more idiomatic. Is there a good reason not to leave this as is?

File: benchmarks/experiment_runner.py [hunk 2]
```diff
@@ -44,97 +44,113 @@ def __init__(self, args):
...
 
   def run(self):
-    if self._args.experiment_config and self._args.model_config:
-      if self._args.dry_run:
-        logger.warning(f"Dry run with {[sys.executable] + sys.argv}")
-        return
-      experiment_config = json.loads(self._args.experiment_config)
-      model_config = json.loads(self._args.model_config)
-      self.run_single_experiment(experiment_config, model_config)
+    is_entry_process = self._args.experiment_config is None and self._args.model_config is None
```
Comment (vanbasten23) [inline benchmarks/experiment_runner.py:47]: by "entry_process", do you mean the main process?

File: benchmarks/experiment_runner.py [hunk 3]
```diff
@@ -44,97 +44,113 @@ def __init__(self, args):
...
-          else:
-            e = "SKIP because of incompatible model and experiment configs."
-            logger.warning(e)
+      assert self._args.experiment_config is not None and self._args.model_config is not None
+      self.run_single_config()
+
+  def generate_and_run_all_configs(self):
+    assert self._args.experiment_config is None and self._args.model_config is None
+
+    # Collect fingerprints for configs to skip. These are configs for which we already have results.
```
Comment (vanbasten23) [inline benchmarks/experiment_runner.py:57]: What's a "fingerprint"?

File: benchmarks/experiment_runner.py [hunk 4]
```diff
@@ -144,19 +160,16 @@ def run_single_experiment(self, experiment_config, model_config):
...
                                              benchmark_model)
         output = move_to_device(output, 'cpu')
         outputs.append(output)
         for key, val in run_metrics.items():
-          # metrics from repeated runs are formed into lists in the metrics dict
           if key not in metrics:
             metrics[key] = []
           metrics[key].append(val)
 
-    # additional experiment metrics can be added here
```
Comment (vanbasten23) [inline benchmarks/experiment_runner.py:158]: Can we leave the comment as is, if it's still valid?

**Comment by vanbasten23** [comment]
Would you mind adding more context about what this PR is trying to achieve in the pr description?

File: benchmarks/util.py (hunk 1 above)
Comment (frgossen) [reply benchmarks/util.py:44]: I'm not sure what is more idiomatic but it is def not the same. `foo` will evaluate to false for a lot of things, e.g. the empty string, an empty list, ... 
With `is not None` the comparison is stricter and less ambiguous to read (?)

File: benchmarks/experiment_runner.py (hunk 2 above)
Comment (frgossen) [reply benchmarks/experiment_runner.py:47]: Yes, changed the name.

File: benchmarks/experiment_runner.py (hunk 3 above)
Comment (frgossen) [reply benchmarks/experiment_runner.py:57]: Added a comment.

File: benchmarks/util.py (hunk 1 above)
Comment (cota) [reply benchmarks/util.py]: Yeah I meant in this case you'll get the same result. Indeed one check is more specific than the other.

FWIW when I see "if foo is not None" or similar, that makes me think that foo might indeed be `0`, `[]`, `{}`, etc. In this case I don't think that can happen, hence my suggestion for what I believe is more idiomatic. But in any case this is a nit.

**Comment by frgossen** [comment]
> Would you mind adding more context about what this PR is trying to achieve in the pr description?

Ofc, added a description.

**Review by cota — APPROVED** [review]

**Pull request merged**